File: UI/stonecarve_manager_flutter/lib/models/material.dart
import 'product.dart';

class StoneMaterial {
  int? id;
  String? name;
  String? description;
  String? imageUrl;
  double? pricePerUnit;
  String? unit;
  int? quantityInStock;
  bool? isAvailable;
  bool? isActive;
  DateTime? createdAt;
  DateTime? updatedAt;

  // Related objects
  List<Product>? products;

  StoneMaterial({
    this.id,
    this.name,
    this.description,
    this.imageUrl,
    this.pricePerUnit,
    this.unit,
    this.quantityInStock,
    this.isAvailable,
    this.isActive,
    this.createdAt,
    this.updatedAt,
    this.products,
  });

  StoneMaterial.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    name = json['name'];
    description = json['description'];
    imageUrl = json['imageUrl'];
    pricePerUnit = json['pricePerUnit']?.toDouble();
    unit = json['unit'];
    quantityInStock = json['quantityInStock'];
    isAvailable = json['isAvailable'];
    isActive = json['isActive'];
    createdAt = json['createdAt'] != null
        ? DateTime.parse(json['createdAt'])
        : null;
    updatedAt = json['updatedAt'] != null
        ? DateTime.parse(json['updatedAt'])
        : null;

    if (json['products'] != null) {
      products = <Product>[];
      json['products'].forEach((v) {
        products!.add(Product.fromJson(v));
      });
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'imageUrl': imageUrl,
      'pricePerUnit': pricePerUnit,
      'unit': unit,
      'quantityInStock': quantityInStock,
      'isAvailable': isAvailable,
      'isActive': isActive,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
      'products': products?.map((v) => v.toJson()).toList(),
    };
  }
}

File: StoneCarveManager.Services/Database/Entities/Material.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace StoneCarveManager.Services.Database.Entities
{
    public class Material
    {
        public int Id { get; set; }
        public string Name { get; set; } = string.Empty; // e.g. Marble, Granite, Limestone

        public string Description { get; set; } = string.Empty;

        public string? ImageUrl { get; set; }

        public decimal PricePerUnit { get; set; } // Price per unit

        public string Unit { get; set; } = "m²"; // m², kg, pcs

        public int QuantityInStock { get; set; } = 0;

        public bool IsAvailable { get; set; } = true;

        public bool IsActive { get; set; } = true;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        public DateTime? UpdatedAt { get; set; }

        // Navigation property
        public ICollection<Product> Products { get; set; } = new List<Product>();
    }
}
